Cache document zone stack and simplify caching strategy in conjunction with Celery.

jezdez opened the pull request.

groovecoder commented on kuma/wiki/jobs.py:
```diff
@@ -0,0 +1,56 @@
+                  for zone in zones]
+        return remaps
+
+    def key(self, locale):
```
Should we make a `KumaJob(Job)` super-class to hold some of the common `key` and `empty` logic? Is there enough identical behavior to make us want to re-use them?

jezdez replied:
Sure why not. FTR, the only reason why I even override the `key` method is because the automatic approach implemented in cacheback doesn't correctly take unicode vs. byte string differences into account. Which this simple key method does.

groovecoder replied:
Maybe send an upstream pull request :question:

jezdez replied:
Did that, I think that solved the problem, https://github.com/codeinthehole/django-cacheback/pull/39

jezdez replied:
In other words, I'm not going to add a `KumaJob` for now since we can rely on the automatic key generation.

groovecoder commented on kuma/wiki/models.py:
```diff
@@ -333,31 +340,40 @@ def regenerate_cache_with_fields(self):
+            content = self.rendered_html
+        else:
+            content = self.html
+        return self.extract_section(content, section_id, ignore_heading)
```
:+1: for more readable code

groovecoder replied:
oh, **and** for re-using code!

groovecoder commented on kuma/wiki/tests/test_models.py:
```diff
@@ -19,18 +19,19 @@
+                          DocumentRenderingInProgress)
+from ..jobs import DocumentZoneStackJob
+from ..models import (Document, Revision, RevisionIP,
+                      DocumentZone, TaggedDocument)
```
nit: could put `DocumentZone` up on the preceding line now?
```
from ..models import (Document, Revision, RevisionIP, DocumentZone,
                      TaggedDocument)
```

jezdez replied:
:+1:

groovecoder commented on tox.ini:
```diff
@@ -7,7 +7,7 @@ deps=
     -rrequirements/dev.txt
 commands =
-    coverage run manage.py test --noinput -v2 kuma
+    coverage run manage.py test --noinput -v2 {posargs:kuma}
```
This change is also in https://github.com/mozilla/kuma/pull/3109/files#diff-b91f3d5bd63fcd17221b267e851608e8 so it may potentially conflict?

jezdez replied:
Nah, git will auto merge that since there is nothing else in the other branch changing on that line.

groovecoder commented:
Code review done ... moving on to spot check.

groovecoder commented:
Found a bug where a new zone isn't displayed as such ...

Steps to reproduce:
1. Make a new document
2. Convert it into a zone
3. Refresh it

Expected result:
The doc should show up as a content zone. (i.e., including the `zone-landing` classes, and the "Manage content zone" link in the gear menu)

Actual result:
The doc displays as a regular doc.

jezdez commented:
Aha, thanks @groovecoder!

jezdez commented:
@groovecoder I've fixed the things you saw, I think -- using the `post_save` signal for documents instead of the `render_done` one. Let me know if it works for you. I did find an issue with cacheback actually, as you suggested and fixed it locally. PR: https://github.com/codeinthehole/django-cacheback/pull/40

groovecoder commented:
Looks good. I was able to make a new doc, convert it to zone, and see it rendered as a zone. I also made a child page of the zone landing page and it immediately rendered as the right zone too.
